### PR TITLE
[CALCITE-6314] Add RANDOM function (enabled in Postgres library)

### PR DESCRIPTION
**Pull request by jduo**

### CODE REVIEW

**Comment by jduo** [comment]
@tanclary

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java [hunk 1]
```diff
@@ -2257,4 +2258,16 @@ private static RelDataType deriveTypeMapFromEntries(SqlOperatorBinding opBinding
   @LibraryOperator(libraries = {SPARK})
   public static final SqlFunction GETBIT =
       BIT_GET.withName("GETBIT");
+
+  /** The RANDOM() function. Equivalent to RAND(). */
+  @LibraryOperator(libraries = {POSTGRESQL})
+  public static final SqlFunction RANDOM = new SqlRandFunction() {
```
Comment (tanclary) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:2264]: what's to keep you from doing RAND.withName("RANDOM").withReturnTypeInferance(...)` I think that's better than these overrides

File: site/_docs/reference.md [hunk 2]
```diff
@@ -2813,6 +2813,7 @@ In the following:
 | b | PARSE_TIMESTAMP(format, string[, timeZone])    | Uses format specified by *format* to convert *string* representation of timestamp to a TIMESTAMP WITH LOCAL TIME ZONE value in *timeZone*
 | h s | PARSE_URL(urlString, partToExtract [, keyToExtract] ) | Returns the specified *partToExtract* from the *urlString*. Valid values for *partToExtract* include HOST, PATH, QUERY, REF, PROTOCOL, AUTHORITY, FILE, and USERINFO. *keyToExtract* specifies which query to extract
 | b s | POW(numeric1, numeric2)                      | Returns *numeric1* raised to the power *numeric2*
+| p | RANDOM()                                       | Generates a random double between 0 and 1 inclusive
```
Comment (tanclary) [inline site/_docs/reference.md:2816]: Is there already documentation for RAND? If so, just say `Equivalent to RAND` (see other functions w multiple names for examples)

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 3]
```diff
@@ -6307,6 +6307,22 @@ void checkRegexpExtract(SqlOperatorFixture f0, FunctionAlias functionAlias) {
...
+    final SqlOperatorFixture f = fixture();
+    f.setFor(SqlLibraryOperators.RANDOM, VmName.EXPAND);
+    f.checkFails("^random^", "Column 'RANDOM' not found in any table", false);
+    Consumer<SqlOperatorFixture> test = fixture -> {
+      for (int i = 0; i < 100; i++) {
+        // Result must always be between 0 and 1, inclusive.
+        fixture.checkScalarApprox("random()", "DOUBLE NOT NULL", isWithin(0.5, 0.5));
+      }
+    };
+    f.forEachLibrary(list(SqlLibrary.POSTGRESQL), test);
```
Comment (tanclary) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:6323]: do any other main dialects used by Calcite support this function? If so, enable them too please

**Review by tanclary — CHANGES_REQUESTED** [review]
Please adjust the JIRA and commit message to follow the style of other function additions, should look something like `[CALCITE-XXXX] Add RANDOM function (enabled in Postgres library)` look through commit history for examples

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java (hunk 1 above)
Comment (jduo) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:2264]: I'd prefer that as well. Those methods are part of SqlBasicFunction, but SqlRandFunction derives directly from SqlFunction.

File: site/_docs/reference.md (hunk 2 above)
Comment (jduo) [reply site/_docs/reference.md:2816]: This is from the RAND documentation, minus the comment about an optional seed. In this case where the functions are not exactly the same due to the missing overload, should we still use "Equivalent to"...?

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java (hunk 3 above)
Comment (jduo) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:6323]: Looks like Redshift is the only other DBMS supporting this.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java (hunk 1 above)
Comment (jduo) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:2264]: Done

Comment (tanclary) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:2264]: can't you make SqlRandFunction extend SqlBasicfunction

File: site/_docs/reference.md (hunk 2 above)
Comment (tanclary) [reply site/_docs/reference.md:2816]: ah sorry didn't see there's a difference, this is fine

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java (hunk 1 above)
Comment (jduo) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:2264]: SqlBasicFunction has a private constructor.
I'll look into making RAND be an instance of SqlBasicFunction then RANDOM being a variant instead.

Comment (jduo) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java]: Done.

Comment (tanclary) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java]: the constructor is private because you're meant to use `.create()`

Comment (jduo) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java]: Yes, I wound up adding more with*() functions to SqlBasicFunction for setting dynamic and the OperandTypeValidator and made both RAND() and RANDOM() instances of it.

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 4]
```diff
@@ -6307,6 +6307,22 @@ void checkRegexpExtract(SqlOperatorFixture f0, FunctionAlias functionAlias) {
...
   }
 
+  /** Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-6314">[CALCITE-6314]
+   * Add RANDOM function (enabled in Postgres library)</a>. */
+  @Test void testRandomFunc() {
+    final SqlOperatorFixture f = fixture();
+    f.setFor(SqlLibraryOperators.RANDOM, VmName.EXPAND);
+    f.checkFails("^random^", "Column 'RANDOM' not found in any table", false);
+    Consumer<SqlOperatorFixture> test = fixture -> {
```
Comment (tanclary) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:6317]: sorry this looks good but can we just call this `consumer` to fit the pattern of other tests? `test` doesn't tell us much

**Review by tanclary — CHANGES_REQUESTED** [review]
One small comment, +1 to merge after

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java (hunk 4 above)
Comment (jduo) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java]: Sure, I've renamed test to consumer.

**Review by tanclary — COMMENTED** [review]
Looks like some merge conflicts?

**Comment by jduo** [comment]
> Looks like some merge conflicts?

Updated the PR. It's introduced from merging CALCITE-6278. I expect we'll see more of this since we have a few devs working on adding functions now.

**Review by tanclary — APPROVED** [review]
LGTM

**Pull request merged**